agent/engine/dockerclient/dockerclientfactory.go

```diff
@@ -121,8 +139,70 @@ func findDockerVersions(endpoint string) map[DockerVersion]dockeriface.Client {
 		err = client.Ping()
 		if err != nil {
 			log.Infof("Failed to ping with Docker version %s: %v", version, err)
+			continue
 		}
 		clients[version] = client
 	}
 	return clients
 }
+
+func findDockerVersionsfromMinMaxVersions(endpoint string) (map[DockerVersion]dockeriface.Client, error) {
+	// get a Docker client with the default supported version
+	client, err := newVersionedClient(endpoint, string(minDockerAPIVersion))
+	if err != nil {
+		return nil, errors.Wrap(err, "min API version check: error while creating default Docker client")
+	}
+
+	clientVersion, err := client.Version()
+	if err != nil {
+		return nil, errors.Wrap(err, "min API version check: error while getting client version")
+	}
+
+	clients := make(map[DockerVersion]dockeriface.Client)
+	// check if the docker.Env obj has MinAPIVersion key
+	if !clientVersion.Exists(minAPIVersionKey) {
+		return nil, errors.New("min API version not found in client version")
+	}
+	minAPIVersion := clientVersion.Get(minAPIVersionKey)
+	apiVersion := clientVersion.Get(apiVersionKey)
+	for _, version := range getKnownAPIVersions() {
+		dockerClient, err := getDockerClientForVersion(endpoint, string(version), minAPIVersion, apiVersion)
+		if err != nil {
+			log.Infof("Unable to get Docker client for version %s: %v", version, err)
+			continue
+		}
+		clients[version] = dockerClient
+	}
+	return clients, nil
+}
+
+func getDockerClientForVersion(
+	endpoint string,
+	version string,
+	minAPIVersion string,
+	apiVersion string) (dockeriface.Client, error) {
+	// Adding patch number zero to Docker versions to reuse the existing semver
+	// comparator
+	// TODO: remove this logic later when non-semver comparator is implemented
+	version += zeroPatch
+	lessThanMinCheck := "<" + minAPIVersion + zeroPatch
+	moreThanMaxCheck := ">" + apiVersion + zeroPatch
+	minVersionCheck, err := utils.Version(version).Matches(lessThanMinCheck)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error while comparing version %s with minAPIVersion %s", version, minAPIVersion)
```

Comment:
Can you rephrase this? `"version detection using MinAPIVersion: unable to get min version: %s"`